proposals/compatibility-mode.md

```diff
@@ -0,0 +1,69 @@
+
+## Problem
+
+WebGPU is a good match for modern explicit graphics APIs such as Vulkan, Metal and D3D12. However, there are a large number of devices which do not yet support those APIs. In particular, on Chrome on Windows, 31% of Chrome users do not have D3D11.1 or higher. On Android, [23% of Android users do not have Vulkan 1.1 (15% do not have Vulkan at all)](https://developer.android.com/about/dashboards). On ChromeOS, Vulkan penetration is still quite low, while OpenGL ES 3.1 is ubiquitous.
+
+## Goals
+
+The primary goal of WebGPU Compatibility mode is to increase the reach of WebGPU by providing an opt-in, slightly restricted subset of WebGPU which will run on older APIs such as D3D11 and OpenGL ES. The set of restrictions in Compatibility mode should be kept to a minimum in order to make it easy to port exsting WebGPU applications. This will increase adoption of WebGPU applications via a wider userbase.
+
+Since WebGPU Compatibility mode is a subset of WebGPU, all valid Compatibility mode applications are also valid WebGPU applications. Consequently, Compatibility mode applications will also run on user agents which do not support Compatibility mode. Such user agents will simply ignore the option requesting a Compatibility mode Adapter and return a Core WebGPU Adapter instead.
+
+## WebGPU Spec Changes
+
+```webidl
+partial dictionary GPURequestAdapterOptions {
+    boolean compatibilityMode = false;
+}
+```
+
+When calling `GPU.RequestAdapter()`, passing `compatibilityMode = true` in the `GPURequestAdapterOptions` will indicate to the User Agent to select the Compatibility subset of WebGPU. Any Devices created from the resulting Adapter on supporting UAs will support only Compatibility mode. Calls to APIs unsupported by Compatibility mode will result in validation errors.
+
+Note that a supporting User Agent may return a `compatibilityMode = true` Adapter which is backed by a fully WebGPU-capable hardware adapter, such as D3D12, Metal or Vulkan, so long as it validates all subsequent API calls made on the Adapter and the objects it vends against the Compatibility subset.
+
+```webidl
+partial interface GPUAdapter {
+    readonly attribute boolean isCompatibilityMode;
+}
+```
+
+As a convenience to the developer, the Adapter returned will have the `isCompatibilityMode` property set to `true`.
+
+
+```webidl
+partial dictionary GPUTextureDescriptor {
+    GPUTextureViewDimension viewDimension;
+}
+```
+
+See "Texture view dimension may be specified", below.
+
+## Compatibility mode restrictions
+
+### 1. Texture view dimension may be specified 
+
+When specifying a texture, a `viewDimension` property determines the views which can be bound from that texture for sampling (see "Proposed IDL changes", above). Binding a view of a different dimension for sampling than specified at texture creation time will cause a validation error. If `viewDimension` is unspecified, use the following algorithm:
```

Comment:
```suggestion
When creating a texture, the `viewDimension` property determines the `dimension` of the view which can be created from that texture (see "Proposed IDL changes", above). Creating a view of a different dimension than specified at texture creation time will cause a validation error. If `viewDimension` is unspecified, use the following algorithm:
```

I'd clarify this is not specific to binding for sampling.

Reply:
This is specific to using a texture as a TEXTURE_BINDING. In particular when you pass a texture view to `createBindGroup` is where this validation would apply.  For a RENDER_ATTACHMENT the viewDimension is not used. Instead, the existing non-compat validation suffice (https://www.w3.org/TR/webgpu/#abstract-opdef-renderable-texture-view)

Reply:
To put it another way, you can create any view you can create in non-compat. It's only when you actually use the view does validation change from non-compat. The validation being specified here only happens on `createBindGroup`  Error if `view.dimension !== texture.viewDimension`

Reply:
I had forgotten it happened in createBindGroup instead of createView. It makes sense though since texture views don't actually exist in GLES.

I wonder if `viewDimension` should be more clearly named? Like `textureBindingViewDimension`...

Reply:
We should change the name if we validate this later but why not validate it earlier (at view creation)? I think it would make things clearer.

Reply:
You can't validate a view creation because then you could not render to a cube map or a 2d array. To render to (to use as a RENDER_ATTACHMENT) requires that you use a '2d' dimension

```js
const pass = encoder.beginRenderPass({
    label: 'render to the positive Z face of a cube map',
    colorAttachments: [
      {
        view: myCubeMapTexture.createView({
          dimension: '2d',
          baseArrayLayer: 4,
          arrayLayerCount: 1,
          baseMipLevel: 0,
          mipLeveLCount: 1,
        }),
        ...
      },
    ],
});
```

vs, use as TEXTURE_BINDING cubemap

```js
    const bindGroup = device.createBindGroup({
      layout: pipeline.getBindGroupLayout(0),
      entries: [
        { binding: 0, resource: sampler },
        { binding: 1, resource: myCubeMapTexture.createView({dimension: 'cube'}) },
      ],
    });
```

Reply:
I see, thanks for the clarification!

Reply:
I've left this as-is for now. Let me know if there are further edits desired here.

Reply:
I still think it should be renamed; I'm ok with @kainino0x suggestion of `textureBindingViewDimension`.